### DISCUSSION TITLE
[nvim-lspconfg] how to disable autoformat for specific file types

### DISCUSSION THREAD

**Discussion by kevclark** [LazyVim/LazyVim#122] [Q&A]
Hi,

Auto format on save is great, however on certain projects I work on I would prefer the option to disable it completely. I realise there is a key binding to temporarily disable this (`<leader>uf`), but it would be ideal if I could do this at local config level.  I would like to do this for any file type which clangd serves, i.e. .c and .h files.

Other than having clangd installed via Mason for diagnostics, I don't have a specific entry for null-ls for clangd.  However `NullLsInfo` does return:

```
       Supported source(s)
	* formatting: astyle | clang_format | uncrustify
	* diagnostics: clang_check | clazy | cppcheck | cpplint | gccdiag
	
	* Note: current buffer has no sources attached  
```
Thanks!

**Comment by folke** [comment]
Just disable formatting in the clangd lsp settings?

LSP servers like clangd also provide formatting functionality, so you need to disable that for clangd. I don't know what setting you need for that. Check the clangd docs

**Reply by kevclark** [reply]
Ah great, I knew that would be a simple answer :smile: I will dig into it.  Thanks!

**Reply by utkarshgupta137** [reply]
Disabling the capability means that I can't format manually either.

**Comment by kevclark** [comment]
Unfortunately formatting setup is not exposed by clangd lsp.
https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#clangd

Although clang-format is not installed via mason, it is in fact embedded into clangd.
https://clangd.llvm.org/features#formatting

The only way to disable it, from the digging I have done, is to add .clang-format files to an existing codebase.  Not ideal, however is this in the only lsp that presents this problem, I guess it doesn't warrant adding a configurable control of `autoformat` based on file type or lsp server.
https://github.com/clangd/clangd/issues/702

**Reply by folke** [reply]
@kevclark https://www.lazyvim.org/configuration/tips#disable-autoformat-for-some-buffers
